File: Indexed Search/Hash Based Indexed Search/Linear Hashing/LinearHashing.java
import java.util.*;

public class LinearHashing {

static int level=0;

static int split=0;

static final int N=4;

static List<
List<Integer>
>
table=
new ArrayList<>();

static{

for(
int i=0;
i<N;
i++
)
table.add(
new ArrayList<>()
);

}

// h(k)=k mod (N×2^L)

static int hash(
int k
){

return
k%
(
N
<<
level
);

}

static void insert(
int key
){

int idx=
hash(
key
);

while(
idx
>=
table.size()
)
table.add(
new ArrayList<>()
);

table
.get(
idx
)
.add(
key
);

}

static boolean search(
int key
){

int idx=
hash(
key
);

return
table
.get(
idx
)
.contains(
key
);

}

public static void main(
String[] args
){

insert(
100
);

insert(
200
);

System.out.println(
search(
200
)
);

}

}

File: Indexed Search/Hash Based Indexed Search/Linear Hashing/LinearHashing.py
class LinearHash:

    def __init__(
            self):

        self.level = 0
        self.split = 0
        self.N = 4

        self.table = [
            []
            for _
            in range(
                self.N
            )
        ]

    # h(k)=k mod (N×2^L)

    def hash(
            self,
            k):

        return (
            k %
            (
                self.N
                <<
                self.level
            )
        )

    def insert(
            self,
            key):

        idx = self.hash(
            key
        )

        while (
                idx
                >=
                len(
                    self.table
                )
        ):
            self.table.append(
                []
            )

        self.table[
            idx
        ].append(
            key
        )

    def search(
            self,
            key):

        idx = self.hash(
            key
        )

        return (
            key
            in
            self.table[
                idx
            ]
        )


h = LinearHash()

h.insert(
10
)

h.insert(
20
)

print(
h.search(
20
)
)
